File: apps/api/src/engram_api/routers/v1/memories.py
"""/memories — typed memory objects: CRUD (as events), the justification spine,
timeline, undo.

Every handler parses input, calls one service, and maps the result (ADR-0007).
Model of record: docs/memory-model.md.
"""

from datetime import datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status

from engram_api.dependencies import (
    get_history_queries,
    get_memory_commands,
    get_memory_queries,
    get_principal,
    get_timeline_queries,
)
from engram_api.schemas.common import PROBLEM_RESPONSES, ProvenanceView
from engram_api.schemas.memories import (
    AddEvidenceRequest,
    AdjustImportanceRequest,
    ConfirmRequest,
    ContradictRequest,
    CreateMemoryRequest,
    EditMemoryRequest,
    EvidenceView,
    LinkView,
    MemoryKindName,
    MemoryListResponse,
    MemoryResponse,
    MemorySnapshotResponse,
    SetLifetimeRequest,
    SetVisibilityRequest,
    TimelineEntryResponse,
    TimelineResponse,
    UndoResponse,
    UpdateAttributesRequest,
)
from engram_core.application.commands.memory_commands import MemoryCommandService
from engram_core.application.dto import (
    CreateMemoryInput,
    EditMemoryInput,
    MemoryReadModel,
    MemorySnapshot,
    UpdateAttributesInput,
)
from engram_core.application.queries.history_queries import HistoryQueryService
from engram_core.application.queries.memory_queries import MemoryQueryService
from engram_core.application.queries.timeline_queries import TimelineQueryService
from engram_core.domain.errors import ValidationError
from engram_core.domain.values import (
    EvidenceRef,
    EvidenceType,
    Lifetime,
    MemoryId,
    MemoryKind,
    RetentionPolicy,
    Visibility,
)
from engram_events import Provenance

router = APIRouter(prefix="/memories", tags=["memories"], responses=PROBLEM_RESPONSES)

Commands = Annotated[MemoryCommandService, Depends(get_memory_commands)]
Queries = Annotated[MemoryQueryService, Depends(get_memory_queries)]
Timeline = Annotated[TimelineQueryService, Depends(get_timeline_queries)]
History = Annotated[HistoryQueryService, Depends(get_history_queries)]
Principal = Annotated[Provenance, Depends(get_principal)]


# -- DTO <-> wire mapping (a field rename, not a transformation; ADR-0021 §1) ----


def _response(model: MemoryReadModel) -> MemoryResponse:
    return MemoryResponse(
        id=UUID(str(model.id)),
        kind=model.kind.value,
        slug=model.slug,
        title=model.title,
        content=model.content,
        attributes=model.attributes,
        tags=list(model.tags),
        links=[
            LinkView(
                target_id=UUID(str(link.target_id)),
                relation=link.relation,  # type: ignore[arg-type]
            )
            for link in model.links
        ],
        evidence=[
            EvidenceView(
                evidence_type=item.evidence_type,  # type: ignore[arg-type]
                value=item.value,
                note=item.note,
                added_at=item.added_at,
                actor=item.actor,
            )
            for item in model.evidence
        ],
        confidence=model.confidence,
        effective_confidence=model.effective_confidence,
        stale=model.stale,
        last_confirmed_at=model.last_confirmed_at,
        lifetime_policy=model.lifetime_policy,  # type: ignore[arg-type]
        lifetime_until=model.lifetime_until,
        visibility=model.visibility,  # type: ignore[arg-type]
        pinned=model.pinned,
        user_weight=model.user_weight,
        archived=model.archived,
        created_at=model.created_at,
        updated_at=model.updated_at,
        version=model.version,
    )


def _snapshot_response(snapshot: MemorySnapshot) -> MemorySnapshotResponse:
    return MemorySnapshotResponse(
        id=UUID(str(snapshot.id)),
        kind=snapshot.kind.value,
        slug=snapshot.slug,
        title=snapshot.title,
        content=snapshot.content,
        attributes=snapshot.attributes,
        tags=list(snapshot.tags),
        confidence=snapshot.confidence,
        lifetime_policy=snapshot.lifetime_policy,  # type: ignore[arg-type]
        visibility=snapshot.visibility,  # type: ignore[arg-type]
        archived=snapshot.archived,
        deleted=snapshot.deleted,
        version=snapshot.version,
    )


def _lifetime(policy: str, until: datetime | None) -> Lifetime:
    return Lifetime(RetentionPolicy(policy), until)


# -- listing & core content -----------------------------------------------------


@router.get("", response_model=MemoryListResponse)
async def list_memories(
    queries: Queries,
    principal: Principal,
    kind: Annotated[MemoryKindName | None, Query()] = None,
    tag: Annotated[str | None, Query()] = None,
    include_archived: Annotated[bool, Query()] = False,
    include_stale: Annotated[bool, Query()] = True,
    cursor: Annotated[str | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=200)] = 50,
) -> MemoryListResponse:
    """Cursor-paginated listing. Visibility enforcement against the principal
    arrives with the auth milestone; today every caller is the owning user."""
    page = queries.list_memories(
        kind=MemoryKind(kind) if kind is not None else None,
        tag=tag,
        include_archived=include_archived,
        include_stale=include_stale,
        cursor=cursor,
        limit=limit,
    )
    return MemoryListResponse(
        items=[_response(m) for m in page.items], next_cursor=page.next_cursor
    )


@router.post("", response_model=MemoryResponse, status_code=status.HTTP_201_CREATED)
async def create_memory(
    body: CreateMemoryRequest, commands: Commands, queries: Queries, principal: Principal
) -> MemoryResponse:
    """Create a typed memory; attributes must satisfy the kind schema (422 otherwise)."""
    memory_id = commands.create_memory(
        CreateMemoryInput(
            kind=MemoryKind(body.kind),
            title=body.title,
            content=body.content,
            attributes=body.attributes,
            slug=body.slug,
            tags=tuple(body.tags),
            confidence=body.confidence,
            lifetime=_lifetime(body.lifetime_policy, body.lifetime_until),
            visibility=Visibility(body.visibility),
        ),
        principal,
    )
    return _response(queries.get_memory(memory_id))


@router.get("/{memory_id}", response_model=MemoryResponse)
async def get_memory(memory_id: UUID, queries: Queries, principal: Principal) -> MemoryResponse:
    """Current state of one memory, spine included."""
    return _response(queries.get_memory(MemoryId(memory_id)))


@router.patch("/{memory_id}", response_model=MemoryResponse)
async def edit_memory(
    memory_id: UUID,
    body: EditMemoryRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Sparse narrative edit (title/content/slug). 409 when ``expected_version`` is stale."""
    identifier = MemoryId(memory_id)
    commands.edit_memory(
        identifier,
        EditMemoryInput(
            expected_version=body.expected_version,
            title=body.title,
            content=body.content,
            slug=body.slug,
        ),
        principal,
    )
    return _response(queries.get_memory(identifier))


@router.patch("/{memory_id}/attributes", response_model=MemoryResponse)
async def update_attributes(
    memory_id: UUID,
    body: UpdateAttributesRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Sparse change to kind-schema fields; validated against the KindRegistry."""
    identifier = MemoryId(memory_id)
    commands.update_attributes(
        identifier,
        UpdateAttributesInput(expected_version=body.expected_version, changes=body.changes),
        principal,
    )
    return _response(queries.get_memory(identifier))


@router.delete("/{memory_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_memory(
    memory_id: UUID,
    commands: Commands,
    principal: Principal,
    reason: Annotated[str | None, Query()] = None,
) -> None:
    """Tombstone the memory (its history remains in the event log). Never called
    by automation (ADR-0011) — there is no proposal-driven path to this route."""
    commands.delete_memory(MemoryId(memory_id), reason, principal)


# -- justification spine ------------------------------------------------------


@router.post("/{memory_id}/confirm", response_model=MemoryResponse)
async def confirm_memory(
    memory_id: UUID,
    body: ConfirmRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Vouch for a memory: raises confidence, resets staleness."""
    identifier = MemoryId(memory_id)
    commands.confirm_memory(identifier, body.note, principal)
    return _response(queries.get_memory(identifier))


@router.post("/{memory_id}/contradict", response_model=MemoryResponse)
async def contradict_memory(
    memory_id: UUID,
    body: ContradictRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Dispute a memory: lowers confidence, creates a ``contradicts`` edge."""
    identifier = MemoryId(memory_id)
    contradicting = MemoryId(body.contradicting_id) if body.contradicting_id is not None else None
    commands.contradict_memory(identifier, contradicting, body.note, principal)
    return _response(queries.get_memory(identifier))


@router.post(
    "/{memory_id}/evidence", response_model=MemoryResponse, status_code=status.HTTP_201_CREATED
)
async def add_evidence(
    memory_id: UUID,
    body: AddEvidenceRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Append supporting evidence (append-only)."""
    identifier = MemoryId(memory_id)
    commands.add_evidence(
        identifier,
        EvidenceRef(EvidenceType(body.evidence_type), body.value, body.note),
        principal,
    )
    return _response(queries.get_memory(identifier))


@router.patch("/{memory_id}/importance", response_model=MemoryResponse)
async def adjust_importance(
    memory_id: UUID,
    body: AdjustImportanceRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Pin/unpin or set the explicit user weight."""
    identifier = MemoryId(memory_id)
    commands.adjust_importance(
        identifier, pinned=body.pinned, user_weight=body.user_weight, provenance=principal
    )
    return _response(queries.get_memory(identifier))


@router.patch("/{memory_id}/visibility", response_model=MemoryResponse)
async def set_visibility(
    memory_id: UUID,
    body: SetVisibilityRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Change who may recall this memory."""
    identifier = MemoryId(memory_id)
    commands.set_visibility(
        identifier, Visibility(body.visibility), tuple(body.allowed_actors), principal
    )
    return _response(queries.get_memory(identifier))


@router.patch("/{memory_id}/lifetime", response_model=MemoryResponse)
async def set_lifetime(
    memory_id: UUID,
    body: SetLifetimeRequest,
    commands: Commands,
    queries: Queries,
    principal: Principal,
) -> MemoryResponse:
    """Change the retention policy."""
    identifier = MemoryId(memory_id)
    commands.set_lifetime(
        identifier, _lifetime(body.lifetime_policy, body.lifetime_until), principal
    )
    return _response(queries.get_memory(identifier))


# -- history --------------------------------------------------------------------


@router.get("/{memory_id}/timeline", response_model=TimelineResponse)
async def memory_timeline(memory_id: UUID, timeline: Timeline) -> TimelineResponse:
    """Full event history of one memory, oldest first — provenance included,
    the same shape the Observatory reads (ADR-0021 §3, ADR-0022)."""
    entries = timeline.memory_timeline(MemoryId(memory_id))
    return TimelineResponse(
        memory_id=memory_id,
        entries=[
            TimelineEntryResponse(
                event_id=entry.event_id,
                event_type=entry.event_type,
                occurred_at=entry.occurred_at,
                stream_seq=entry.stream_seq,
                provenance=ProvenanceView(
                    actor=entry.actor, session_id=entry.session_id, detail=entry.detail
                ),
            )
            for entry in entries
        ],
    )


@router.get("/{memory_id}/at", response_model=MemorySnapshotResponse)
async def memory_at(
    memory_id: UUID,
    history: History,
    at: Annotated[datetime | None, Query(description="Instant to reconstruct (ISO-8601)")] = None,
    version: Annotated[int | None, Query(ge=1, description="Stream version to reconstruct")] = None,
) -> MemorySnapshotResponse:
    """Time travel (ADR-0002/ADR-0021): the memory exactly as it was at a moment
    or version. Exactly one of ``at``/``version`` is required (422 otherwise)."""
    if (at is None) == (version is None):
        raise ValidationError("time travel needs exactly one of ?at= or ?version=")
    snapshot = history.state_at(MemoryId(memory_id), at=at, version=version)
    return _snapshot_response(snapshot)


@router.post("/{memory_id}/undo", response_model=UndoResponse)
async def undo_last_change(
    memory_id: UUID, commands: Commands, queries: Queries, principal: Principal
) -> UndoResponse:
    """Append the compensating event for the most recent change on this memory
    (ADR-0021). 409 when the last event has no defined inverse."""
    identifier = MemoryId(memory_id)
    compensating_id = commands.undo_last_change(identifier, principal)
    return UndoResponse(
        memory=_response(queries.get_memory(identifier)), compensating_event_id=compensating_id
    )

File: apps/web/src/lib/api/hooks/memories.ts
"use client";

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { apiClient } from "@/lib/api/client";
import { unwrap } from "@/lib/api/errors";
import { type MemoryListFilters, queryKeys } from "@/lib/api/keys";

/**
 * Memory Explorer's data layer. Every function here is parse-params →
 * one `@engram/api-client` call → typed result; no folding, no derived
 * confidence, no decisions — those all happen server-side (ADR-0021 §1).
 */

export function useMemories(filters: MemoryListFilters = {}) {
  return useQuery({
    queryKey: queryKeys.memories(filters),
    queryFn: async () => {
      const result = await apiClient.GET("/api/v1/memories", {
        params: {
          query: {
            kind: filters.kind as never,
            tag: filters.tag,
            include_archived: filters.includeArchived,
            include_stale: filters.includeStale,
            cursor: filters.cursor,
          },
        },
      });
      return unwrap(result);
    },
  });
}

export function useMemory(memoryId: string | undefined) {
  const id = memoryId ?? "";
  return useQuery({
    queryKey: queryKeys.memory(id),
    // `enabled` gates whether this ever runs — an empty id here never
    // reaches the network, so a fallback is safe and avoids `memoryId!`.
    queryFn: async () => {
      const result = await apiClient.GET("/api/v1/memories/{memory_id}", {
        params: { path: { memory_id: id } },
      });
      return unwrap(result);
    },
    enabled: memoryId !== undefined,
  });
}

export function useMemoryTimeline(memoryId: string | undefined) {
  const id = memoryId ?? "";
  return useQuery({
    queryKey: queryKeys.memoryTimeline(id),
    queryFn: async () => {
      const result = await apiClient.GET("/api/v1/memories/{memory_id}/timeline", {
        params: { path: { memory_id: id } },
      });
      return unwrap(result);
    },
    enabled: memoryId !== undefined,
  });
}

/** Time travel (ADR-0002/ADR-0021): exactly one of version/at, chosen by the caller. */
export function useMemoryAt(
  memoryId: string | undefined,
  selector: { version?: number; at?: string },
  options?: { enabled?: boolean },
) {
  const id = memoryId ?? "";
  return useQuery({
    queryKey: queryKeys.memoryAt(id, selector),
    queryFn: async () => {
      const result = await apiClient.GET("/api/v1/memories/{memory_id}/at", {
        params: {
          path: { memory_id: id },
          query: { version: selector.version, at: selector.at },
        },
      });
      return unwrap(result);
    },
    enabled:
      memoryId !== undefined &&
      (selector.version !== undefined || selector.at !== undefined) &&
      (options?.enabled ?? true),
  });
}

function useMemoryMutation<TArgs extends { memoryId: string }, TData>(
  mutationFn: (args: TArgs) => Promise<TData>,
) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn,
    onSuccess: (_data, variables) => {
      void queryClient.invalidateQueries({ queryKey: queryKeys.memory(variables.memoryId) });
      void queryClient.invalidateQueries({ queryKey: queryKeys.memories() });
    },
  });
}

/** Vouch for a memory: raises confidence, resets staleness (the spine's positive signal). */
export function useConfirmMemory() {
  return useMemoryMutation(async ({ memoryId, note }: { memoryId: string; note?: string }) => {
    const result = await apiClient.POST("/api/v1/memories/{memory_id}/confirm", {
      params: { path: { memory_id: memoryId } },
      body: { note: note ?? null },
    });
    return unwrap(result);
  });
}

/** Dispute a memory: lowers confidence (the spine's negative signal). */
export function useContradictMemory() {
  return useMemoryMutation(
    async ({
      memoryId,
      contradictingId,
      note,
    }: {
      memoryId: string;
      contradictingId?: string;
      note?: string;
    }) => {
      const result = await apiClient.POST("/api/v1/memories/{memory_id}/contradict", {
        params: { path: { memory_id: memoryId } },
        body: { contradicting_id: contradictingId ?? null, note: note ?? null },
      });
      return unwrap(result);
    },
  );
}

/** Append supporting evidence — additive, never edits or removes prior evidence. */
export function useAddEvidence() {
  return useMemoryMutation(
    async ({
      memoryId,
      evidenceType,
      value,
      note,
    }: {
      memoryId: string;
      evidenceType: string;
      value: string;
      note?: string;
    }) => {
      const result = await apiClient.POST("/api/v1/memories/{memory_id}/evidence", {
        params: { path: { memory_id: memoryId } },
        body: { evidence_type: evidenceType as never, value, note: note ?? null },
      });
      return unwrap(result);
    },
  );
}

/** Pin/unpin, or set an explicit weight — the spine's importance half. */
export function useAdjustImportance() {
  return useMemoryMutation(
    async ({
      memoryId,
      pinned,
      userWeight,
    }: {
      memoryId: string;
      pinned?: boolean;
      userWeight?: number;
    }) => {
      const result = await apiClient.PATCH("/api/v1/memories/{memory_id}/importance", {
        params: { path: { memory_id: memoryId } },
        body: { pinned: pinned ?? null, user_weight: userWeight ?? null },
      });
      return unwrap(result);
    },
  );
}

/** Compensate the memory's most recent change (ADR-0021). 409 if it has no inverse. */
export function useUndoMemory() {
  return useMemoryMutation(async ({ memoryId }: { memoryId: string }) => {
    const result = await apiClient.POST("/api/v1/memories/{memory_id}/undo", {
      params: { path: { memory_id: memoryId } },
    });
    return unwrap(result);
  });
}
